src/group_theory/burnside.lean

```diff
@@ -0,0 +1,89 @@
+
+variables (G : Type u) [group G] (α : Type v) [mul_action G α]
+
+variables G α
+
+local attribute [instance] mul_action.orbit_rel
+
+/-- A trivial equivalence used in the proof of Burnside Lemma -/
+def sigma_fixed_points_equiv_sigma_stabilizer :
+  (Σ g : G, (to_perm G α g).fixed_points) ≃
+    (Σ (o : orbits G α) (x : {x // ⟦x⟧ = o}), stabilizer G x.val) :=
+{ to_fun  := λ y, ⟨⟦y.snd.val⟧, ⟨⟨y.snd.val, rfl⟩, ⟨y.fst, y.snd.property⟩⟩⟩,
+  inv_fun := λ y, ⟨y.snd.snd.val, ⟨y.snd.fst.val, y.snd.snd.property⟩⟩,
+  left_inv := λ ⟨g, ⟨x, h⟩⟩, rfl,
+  right_inv := λ ⟨o, ⟨⟨x, hx⟩, ⟨g, h⟩⟩⟩, by subst o }
+
+open classical
+
+variables {G α}
+
+/-- Choose an element of the group that sends `o.out` to a specific point of the orbit. -/
+noncomputable def orbit_choice {o : orbits G α} (x : {x // ⟦x⟧ = o}) : {y // y • o.out = x.val} :=
+  indefinite_description _ (quotient.exact $ x.property.trans o.out_eq.symm)
+
+/-- Disjoint union of stabilizers of all points of a given orbit has the same cardinality as the group. -/
+noncomputable def orbit_sigma_stabilizer_equiv_group (o : orbits G α) :
+  (Σ x : {x // ⟦x⟧ = o}, stabilizer G x.val) ≃ G :=
+begin
+  have h₁ : (Σ x : {x // ⟦x⟧ = o}, stabilizer G x.val) ≃
+    ({x // ⟦x⟧ = o} × (stabilizer G o.out)),
+    by { apply equiv.sigma_equiv_prod_of_equiv,
+         assume x,
+         rw [← (orbit_choice x).2],
+         symmetry,
+         apply stabilizer_equiv_smul },
+
+  calc _ ≃ ({x // ⟦x⟧ = o} × (stabilizer G o.out))  : h₁
+    ...  ≃ ((orbit G o.out) × (stabilizer G o.out)) : equiv.prod_congr
+                                                        (equiv.subtype_congr_right $ o.out_eq_iff)
+                                                        (equiv.refl _)
+    ...  ≃ G                                        : orbit_prod_stabilizer_equiv_group G o.out
+end
+
+/-- Equivalence behind Burnside Lemma: disjoint union of fixed points
+    of all elements of the group is equivalent to the product of the
+    group and the orbits space. -/
+noncomputable def sigma_fixed_points_equiv_group_prod_orbits :
+  (Σ g : G, (to_perm G α g).fixed_points) ≃ (G × orbits G α) :=
+calc _ ≃ (Σ (o : orbits G α) (x : {x // ⟦x⟧ = o}), stabilizer G x.val) : sigma_fixed_points_equiv_sigma_stabilizer G α
+   ... ≃ (Σ (o : orbits G α), G)                                       : equiv.sigma_congr_right orbit_sigma_stabilizer_equiv_group
+   ... ≃ (orbits G α × G)                                              : equiv.sigma_equiv_prod _ _
+   ... ≃ (G × orbits G α)                                              : equiv.prod_comm _ _
+
+open fintype finset
+
+variables [fintype G] [decidable_eq G] [fintype α] [decidable_eq α]
+
+/-- Burnside's lemma: number of orbits of a group action equals the average number of fixed points of a group element.
+    For the case of infinite sets, see `burnside_equiv`. -/
+lemma card_group_mul_card_orbits_eq : card G * card (orbits G α) = sum univ (λ g : G, card $ (to_perm G α g).fixed_points) :=
```

Comment:
The `fixed_points` in this definition is not the same as `mul_action.fixed_points`, and this could cause confusion. I think the name of `mul_action.fixed_points` should be changed.